### PR TITLE
Storage method agnostic object storage

### PR DESCRIPTION
**Pull request by kenoir**
### What is this PR trying to achieve?

Allow the VHS to be storage type agnostic.

### Who is this change for?

😭 🍇

### CODE REVIEW

File: catalogue_pipeline/transformer/src/main/scala/uk/ac/wellcome/platform/transformer/receive/SQSMessageReceiver.scala [hunk 1]
```diff
@@ -73,11 +73,11 @@ class SQSMessageReceiver @Inject()(
...
 
   val miroTransformableStore =
-    new S3TypedObjectStore[MiroTransformable](s3StringStore)
+    new S3TypeStore[MiroTransformable](s3StringStore)
   val calmTransformableStore =
-    new S3TypedObjectStore[CalmTransformable](s3StringStore)
+    new S3TypeStore[CalmTransformable](s3StringStore)
   val sierraTransformableStore =
-    new S3TypedObjectStore[SierraTransformable](s3StringStore)
+    new S3TypeStore[SierraTransformable](s3StringStore)
```
Comment (alexwlchan) [inline catalogue_pipeline/transformer/src/main/scala/uk/ac/wellcome/platform/transformer/receive/SQSMessageReceiver.scala]: FYI, I’ll push this naming back into my branch – there’s no point introducing it in diff 1 and immediately changing it in diff 2. Should make this PR a lot smaller too.

File: sbt_common/storage/src/main/scala/uk/ac/wellcome/storage/s3/S3TypeStore.scala [hunk 2]
```diff
@@ -1,26 +1,37 @@
...
 import scala.concurrent.Future
+import scala.io.Source
+
 
 class S3TypeStore[T] @Inject()(
-  stringStore: S3StringStore
-) extends Logging {
+    s3Client: AmazonS3,
+    s3Config: S3Config
+  )(implicit encoder: Encoder[T], decoder: Decoder[T])
```
Comment (alexwlchan) [inline sbt_common/storage/src/main/scala/uk/ac/wellcome/storage/s3/S3TypeStore.scala]: I feel a bit iffy about this – if I’m creating a read-only instance of the TypeStore, why do I also need to supply an encoder?

File: sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/test/fixtures/LocalVersionedHybridStore.scala [hunk 3]
```diff
@@ -62,12 +126,25 @@ trait LocalVersionedHybridStore
...
+
+  def getContentFor(bucket: Bucket, table: Table, id: String) = {
+    val hybridRecord = getHybridRecord(bucket, table, id)
+
+    getContentFromS3(bucket, hybridRecord.s3key)
+  }
+
+  private def getHybridRecord(bucket: Bucket, table: Table, id: String) =
+    Scanamo.get[HybridRecord](dynamoDbClient)(table.name)('id -> id) match {
+      case None => throw new RuntimeException(s"No object with id $id found!")
```
Comment (alicefuzier) [inline sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/test/fixtures/LocalVersionedHybridStore.scala]: I'd rather leave it as it was with the .gets. This hides the original error thrown by Scanamo and I think in tests you want to see it

File: sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/vhs/StreamStoreVersionedHybridStoreTest.scala [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+          val inputStream = new ByteArrayInputStream(content.getBytes)
+
+          val future = hybridStore.updateRecord(id)(inputStream)(identity)()
+
+          whenReady(future) { _ =>
+            getContentFor(bucket, table, id) shouldBe content
+          }
+      }
+    }
+  }
```
Comment (alicefuzier) [inline sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/vhs/StreamStoreVersionedHybridStoreTest.scala]: Could we add a test for the get method as well?

File: sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/vhs/StringStoreVersionedHybridStoreTest.scala [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+              .updateRecord(id)(record)(identity)()
+              .flatMap(_ => hybridStore.updateRecord(id)(record)(t)())
+
+          whenReady(future) { _ =>
+            getContentFor(bucket, table, id) shouldBe expectedRecord
+          }
+      }
+    }
+
+    it("updates DynamoDB and S3 if it sees a new version of a record") {
```
Comment (alicefuzier) [inline sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/vhs/StringStoreVersionedHybridStoreTest.scala]: Isn't this the same test case as above? And I don't see any assertion about the Dynamo content

File: sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/test/fixtures/LocalVersionedHybridStore.scala (hunk 3 above)
Comment (kenoir) [reply sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/test/fixtures/LocalVersionedHybridStore.scala]: I found:
```
Scanamo
      .get[HybridRecord](dynamoDbClient)(table.name)('id -> record.id)
      .get
      .right
      .get
```
To hide the location of the error quite effectively before this change. The error is surfaced in the test output via: `RuntimeException(s"Error reading from dynamo: $error")`

File: sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/vhs/StringStoreVersionedHybridStoreTest.scala (hunk 5 above)
Comment (kenoir) [reply sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/vhs/StringStoreVersionedHybridStoreTest.scala]: yup, noted.

File: sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/vhs/StreamStoreVersionedHybridStoreTest.scala (hunk 4 above)
Comment (kenoir) [reply sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/vhs/StreamStoreVersionedHybridStoreTest.scala]: yup, will do.

File: sbt_common/messaging/src/main/scala/uk/ac/wellcome/messaging/message/MessageReader.scala [hunk 6]
```diff
@@ -19,21 +19,16 @@ class MessageReader[T] @Inject()(
   messageReaderConfig: MessageReaderConfig,
   s3Client: AmazonS3,
   sqsClient: AmazonSQS
-) {
+)(implicit encoder: Encoder[T], decoder: Decoder[T]) {
```
[the remark this answers is not in the document]
Comment (kenoir) [reply sbt_common/messaging/src/main/scala/uk/ac/wellcome/messaging/message/MessageReader.scala]: A bit.

File: sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/test/fixtures/LocalVersionedHybridStore.scala (hunk 3 above)
Comment (alicefuzier) [reply sbt_common/storage/src/test/scala/uk/ac/wellcome/storage/test/fixtures/LocalVersionedHybridStore.scala]: Ah I see

**Pull request merged**